File: haskell/word-count/WordCount.hs
{-
Copyright 2014 Bas Bossink (bas.bossink@gmail.com).
-}
module WordCount (wordCount) where

import Data.Char (isAlphaNum, toLower)
import Data.Foldable (foldl')
import Data.List.Split (wordsBy)
import Data.Map.Strict (Map, fromListWith)

tallyWords :: [String] -> Map String Integer
tallyWords words = fromListWith (+) pairs
  where
  lowerWord = map toLower
  loweredWords = map lowerWord words
  pairs = zip loweredWords $ repeat 1

wordCount :: String -> Map String Integer
wordCount = tallyWords . wordsBy (not . isAlphaNum)

File: javascript/word-count/word-count.js
/*
 * Copyright 2015 Bas Bossink <bas.bossink@gmail.com>
 * This code is licensed under the GNU General Public License version 3.
 */
(function () {
  'use strict';
  var words = function (input) {
    var split = input.split(/\s+/), counts = {}, word, i = 0,
      numberOfWords = split.length;
    for (i = 0; i < numberOfWords; i += 1) {
      word = split[i];
      if (Object.prototype.hasOwnProperty.call(counts, word)) {
        counts[word] += 1;
      } else {
        counts[word] = 1;
      }
    }
    return counts;
  };
  module.exports = words;
}());
